File: src/main/java/core/procedures/math/trigonometry/Sin.java
package core.procedures.math.trigonometry;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.procedures.math.Multiplication;
import core.scm.BigComplex;
import core.scm.BigRational;
import core.utils.Utils;

import java.math.BigDecimal;
import java.math.BigInteger;

public final class Sin extends AFn {

  public Sin() {
    super(new FnArgsBuilder().min(1).max(1).mandatory(new Class[]{Number.class}).build());
  }

  @Override
  public boolean isPure() {
    return true;
  }

  @Override
  public String getName() {
    return "sin";
  }

  @Override
  public Number apply1(Object arg) {
    return sin((Number)arg);
  }

  public static Number sin(Number number) {
    /* Special cases */
    if (Utils.isZero(number)) {
      return 0L;
    }
    if (number instanceof BigDecimal) {
      return sin((BigDecimal) number);
    } else if (number instanceof BigInteger) {
      return sin((BigInteger)number);
    } else if (number instanceof BigComplex) {
      return sin((BigComplex)number);
    } else if (number instanceof BigRational) {
      return sin(((BigRational)number).toBigDecimal());
    }
    return Math.sin(number.doubleValue());
  }

  public static double sin(BigDecimal bd) {
    double v = bd.doubleValue();
    if (Double.isInfinite(v) || Double.isNaN(v)) {
      return Double.NaN;
    } else {
      return Math.sin(v);
    }
  }

  public static double sin(BigInteger bi) {
    double v = bi.doubleValue();
    if (Double.isInfinite(v) || Double.isNaN(v)) {
      return Double.NaN;
    } else {
      return Math.sin(v);
    }
  }

  public static BigComplex sin(BigComplex c) {
    BigDecimal re = c.getRe();
    BigDecimal im = c.getIm();
    return new BigComplex(Multiplication.apply(Sin.sin(re), Cosh.cosh(im)),
                          Multiplication.apply(Cos.cos(re), Sinh.sinh(im)));
  }
}

File: src/main/java/core/procedures/math/trigonometry/Sin.kt
package core.procedures.math.trigonometry

import core.procedures.AFn
import core.procedures.math.Multiplication
import core.scm.BigComplex
import core.utils.Utils
import kotlin.math.cos
import kotlin.math.cosh
import kotlin.math.sin
import kotlin.math.sinh

class Sin : AFn<Number?, Number>(name = "sin", isPure = true, minArgs = 1, maxArgs = 1,
                                 mandatoryArgsTypes = arrayOf(Number::class.java)) {

    override operator fun invoke(arg: Number?) = when {
        Utils.isZero(arg!!) -> 0L
        arg is BigComplex   -> sin(arg)
        else                -> sin(arg.toDouble())
    }

    companion object {

        private val multiplication = Multiplication()

        fun sin(c: BigComplex): BigComplex {
            val x = c.re.toDouble()
            val y = c.im.toDouble()
            return BigComplex(multiplication(sin(x), cosh(y)), multiplication(cos(x), sinh(y)))
        }
    }
}
